### PR TITLE
[WIP] Add alias for GatherV2

### PR DESCRIPTION
**Pull request by yd8534976**
- Change lmnet as antonio's `lm_fyolo_160_1x1_exp10.py`
- Add operator mapping for `GatherV2`
- Remove quantize_first_convolution option
- Fix `Makefile.tpl`

## How has this been tested?

about fps 20

https://gist.github.com/yd8534976/d56fade43d8048b8702ff0e7477bdd8f

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by iizukak** [comment]
@antonionevado Can you check this PR?

**Comment by iizukak** [comment]
run test

**Comment by hadusam** [comment]
run test

**Comment by antonionevado** [comment]
Oh no, no...
But there are compilation errors!
```
tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:1124:89: error: no matching function for call to ‘func_ConcatOnDepth(const TensorView<float, (MemoryLayout)7> [5], T_UINT [5], int, TensorView<float, (MemoryLayout)7>&)’
 func_ConcatOnDepth(block_7_merged_inputs, block_7_merged_inputs_depth, 5, block_7_merged);
                                                                                         ^
In file included from /tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:25:0:
/tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/include/func/concat_on_depth.h:23:6: note: candidate: template<class T> void func_ConcatOnDepth(const TensorView<T, (MemoryLayout)13>*, T_UINT*, T_UINT, const TensorView<T, (MemoryLayout)13>&)
 void func_ConcatOnDepth(const TensorView<T, MemoryLayout::ChHWBCl> inputs[],
      ^
/tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/include/func/concat_on_depth.h:23:6: note:   template argument deduction/substitution failed:
/tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:1124:89: note:   template argument ‘(MemoryLayout)7’ does not match ‘#‘integer_cst’ not supported by dump_decl#<declaration error>’
 func_ConcatOnDepth(block_7_merged_inputs, block_7_merged_inputs_depth, 5, block_7_merged);
                                                                                         ^
/tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:1124:89: note:   ‘const TensorView<float, (MemoryLayout)7>’ is not derived from ‘const TensorView<T, (MemoryLayout)13>’
make[3]: *** [CMakeFiles/lib.dir/src/network.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
```

**Comment by yd8534976** [comment]
> Oh no, no...
> But there are compilation errors!
> 
> ```
> tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:1124:89: error: no matching function for call to ‘func_ConcatOnDepth(const TensorView<float, (MemoryLayout)7> [5], T_UINT [5], int, TensorView<float, (MemoryLayout)7>&)’
>  func_ConcatOnDepth(block_7_merged_inputs, block_7_merged_inputs_depth, 5, block_7_merged);
>                                                                                          ^
> In file included from /tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:25:0:
> /tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/include/func/concat_on_depth.h:23:6: note: candidate: template<class T> void func_ConcatOnDepth(const TensorView<T, (MemoryLayout)13>*, T_UINT*, T_UINT, const TensorView<T, (MemoryLayout)13>&)
>  void func_ConcatOnDepth(const TensorView<T, MemoryLayout::ChHWBCl> inputs[],
>       ^
> /tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/include/func/concat_on_depth.h:23:6: note:   template argument deduction/substitution failed:
> /tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:1124:89: note:   template argument ‘(MemoryLayout)7’ does not match ‘#‘integer_cst’ not supported by dump_decl#<declaration error>’
>  func_ConcatOnDepth(block_7_merged_inputs, block_7_merged_inputs_depth, 5, block_7_merged);
>                                                                                          ^
> /tmp/test-dlk--TestCodeGeneration-201906120145-3pijjt92/10_det_fyolo_quantize_4_v4_x86_64/code_generation.prj/src/network.cpp:1124:89: note:   ‘const TensorView<float, (MemoryLayout)7>’ is not derived from ‘const TensorView<T, (MemoryLayout)13>’
> make[3]: *** [CMakeFiles/lib.dir/src/network.cpp.o] Error 1
> make[3]: *** Waiting for unfinished jobs....
> ```

That should be legacy issues in tca_support branch. It looks same as https://github.com/blue-oil/blueoil/pull/257  🤔

**Comment by ruimashita** [comment]
@yd8534976 
Where do you use `gather` operation ? Why do we need to add `Gather` ?

**Comment by yd8534976** [comment]
@ruimashita This new lmnet file is from antonio. I think there is a `gather` in `tf.contrib.layers.embedding_lookup_unique()`.

**Comment by antonionevado** [comment]
@ruimashita 
tf.contrib.layers.embedding_lookup_unique() uses Gather internally. We need to import the graph first and then, during optimization, all the nodes we use for pixel embedding are merged into a single node (Lookup node). So, in this case, the Gather node is imported and then optimized away.

**Comment by yd8534976** [comment]
Something wrong...
Screen Shot 2019-06-13 at 19 21 08

**Comment by iizukak** [comment]
@yd8534976 Please update branch and re run test.

**Comment by yd8534976** [comment]
run test

File: lmnet/lmnet/networks/object_detection/lm_fyolo.py [hunk 1]
```diff
@@ -81,11 +94,14 @@ def base(self, images, is_training):
         darknet_block = partial(darknet, is_training=is_training,
                                 activation=self.activation, data_format=self.data_format)
 
-        x = darknet_block("block_1", self.inputs, filters=32, kernel_size=1)
+        x = self.encode_image(self.inputs)
```
Comment (iizukak) [inline lmnet/lmnet/networks/object_detection/lm_fyolo.py]: @yd8534976 
Blueoil has Quantize First Layer option in `blueoil.sh init` command.
Is it possible if Quantize First Layer is `true` then these code are enable,
If Quantize First Layer is `false` then these code are disable?

**Comment by iizukak** [comment]
run test

**Comment by yd8534976** [comment]
@antonionevado
I'm confused why did you use `tf.contrib.layers.embedding_lookup_unique()` instead of `tf.nn.embedding_lookup()` directly. [embedding_lookup_unique](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embedding_lookup_unique) claims that unique table can save communication. I understand there are too large `embedding_granularity` for word embedding. But here, we only have `embedding_granularity` 255, I think unique table may not save communication here. For a simple and clean graph, I think we should use `tf.nn.embedding_lookup()` directly. But I find you have finished corresponding `pass_lookup` already. I'm not sure need we change this part?

**Comment by tkng** [comment]
> I'm confused why did you use tf.contrib.layers.embedding_lookup_unique() instead of tf.nn.embedding_lookup() directly.

Sorry for confusion, because training with `tf.contrib.layers.embedding_lookup_unique() ` is much faster than `tf.nn.embedding_lookup()`. There's no other special reason...

**Comment by iizukak** [comment]
run blueoil test

**Comment by antonionevado** [comment]
@yd8534976 Actually I didn't write the Tensorflow code for `Lookup`. I got it from the experimental code of @tkng and @joelN123 
Yes, could be improved and merged into a single node. This was only for a demo. Take in account the comments of @tkng about training speed before change it.

**Comment by yd8534976** [comment]
```
def embedding_lookup_unique(params, ids, partition_strategy="mod", name=None):
  with ops.name_scope(name, "EmbeddingLookupUnique", [params, ids]):
    ids = ops.convert_to_tensor(ids)
    shape = array_ops.shape(ids)
    ids_flat = array_ops.reshape(
        ids, math_ops.reduce_prod(shape, keepdims=True))
    unique_ids, idx = array_ops.unique(ids_flat)
    unique_embeddings = embedding_ops.embedding_lookup(params, unique_ids,
                                                       partition_strategy)
    embeds_flat = array_ops.gather(unique_embeddings, idx)
    embed_shape = array_ops.concat(
        [shape, array_ops.shape(unique_embeddings)[1:]], 0)
    embeds = array_ops.reshape(embeds_flat, embed_shape)
    embeds.set_shape(ids.get_shape().concatenate(
        unique_embeddings.get_shape()[1:]))
    return embeds
```
@tkng @antonionevado Thanks for comments!
In my understanding, `unique_embeddings` is a subset of `params`. If `unique_embeddings` is much smaller than `params`. It will be faster. I thought image could cover almost all pixel values... So size of `unique_embeddings` will be similar to `params`.. So we could implement this directly by something like `embeds_flat = array_ops.gather(params, ids_flat)`. But I have not tested it yet... If you have tested that it is faster... Maybe I'm wrong....

**Comment by yd8534976** [comment]
@ruimashita 
After embedding, I think old version of `quantize_first_convolution` won't make sense anymore. Is it better to delete this config hyperparameter and change quantization part directly?

**Comment by ruimashita** [comment]
@yd8534976 Yes!! I think so, no need `quantize_first_convolution` anymore.

**Comment by yd8534976** [comment]
run test

**Comment by yd8534976** [comment]
run test

**Comment by iizukak** [comment]
@ruimashita 
@yd8534976 remove all `quantize_first_convolution ` option.
It's your requested change?
Old first layer quantize implementation is not needed anymore, but there is still `quantize_first_convolution` option is needed for option.
My understand is wrong?

**Comment by yd8534976** [comment]
run test

**Comment by ruimashita** [comment]
In current FYolo, user can’t convert model when first quantization option is  ‘no’ .
There are some implementation to fix this issue, 
The one is to delete the option from CLI and always use first quantization is ‘yes’ in FYolo,  for user.
But it should be take care in classification and segmentation, these need to the option.
And for developer, to compare models, need function of quantize first layer

**Comment by yd8534976** [comment]
@ruimashita 

In my understand, If we apply embedding on other tasks. #323 , `quantize_first_convolution false` makes the use of embedding meaningless. I think embedding and `quantize_first_convolution true` should be a set to achieve speed up purpose.
And I'm not sure if user could really understand the difference between `quantize_first_convolution true` and `quantize_first_convolution false`. So I think we should decide for them and only provide our best version of models.

**Comment by yd8534976** [comment]
run test

**Pull request closed without merging**